client/allocrunner/alloc_runner.go

```diff
@@ -579,12 +611,47 @@ func (ar *allocRunner) AllocState() *state.State {
 	return state
 }
 
-// Update the running allocation with a new version received from the server.
-//
+// Update asyncronously updates the running allocation with a new version
+// received from the server.
+// When processing a new update, we will first attempt to drain stale updates
+// from the queue, before appending the new one.
+func (ar *allocRunner) Update(update *structs.Allocation) {
+	select {
+	// Drain queued update from the channel if possible, and check the modify
+	// index
+	case oldUpdate := <-ar.allocUpdatedCh:
+		// If the old update is newer than the replacement, then skip the new one
+		// and return. This case shouldn't happen, but may in the case of a bug
+		// elsewhere inside the system.
+		if oldUpdate.AllocModifyIndex > update.AllocModifyIndex {
+			ar.logger.Warn("Discarding allocation update due to newer alloc revision in queue",
+				"old_modify_index", oldUpdate.AllocModifyIndex,
+				"new_modify_index", update.AllocModifyIndex)
+			ar.allocUpdatedCh <- oldUpdate
+			return
+		}
+	default:
+	}
+
+	// Queue the new update
+	ar.allocUpdatedCh <- update
```

Comment:
We already pop previous queued updates from the queue here, but I think there are maybe two subtle bugs:

1) is a possibility of two Update calls timed as such that they could _both_ push to the queue without a read (given this is only called from a single routine, I'm not sure that this is ever going to happen, but 💻 )
2) I think that actually the handleAllocUpdates routine probably shouldn't stop until we close `destroyCh`?

Reply:
> We already pop previous queued updates from the queue here

Oh! Right! It's amazing how quickly I forgot what happened 10 lines up!

> is a possibility of two Update calls timed as such that they could both push to the queue without a read (given this is only called from a single routine, I'm not sure that this is ever going to happen, but computer )

For a given AR, Update should only be called serially by Client. This ensures proper ordering and removes the race between applying updates out of order. Since actually handling the update is done in a goroutine the Update method is "fast" (never blocks, no IO, no loops, low number of operations), calling Updates serially should be sufficient. (If Updates are happening so quickly there'd be a benefit to parallelizing a simple chan recv/send, then something is probably horribly wrong!)

> I think that actually the handleAllocUpdates routine probably shouldn't stop until we close destroyCh?
 
We did this in the old client and it lead to confusing unactionable log messages: https://github.com/hashicorp/nomad/blob/v0.8.6/client/alloc_runner.go#L1073-L1090

However I *think* the behavior is correct since it's entirely possible an allocation may terminate locally concurrent with the server sending it an update.

So one way or the other we need to drop updates to terminal (waitCh has closed) allocations, but with 2 caveats:

1. Don't try to apply the update. This could lead to weird behavior like the alloc health checking hook coming back to life for a terminal alloc (although there should be guards in place to prevent that).

2. Log at TRACE level.